Switch pytest integration to hook

mfocko opened the pull request:
Switch the pytest integration to the hook instead of wrapping the runner manually.

Fixes #152

TODO:
- [x] Adjust specfile for Fedora, cause apparently the pytest module is not included.
  - Decided to use the specfile directly from the Fedora, since the one, here, is not being kept up to date.
- [x] Rebase
- [x] Test against 8.1.1
- [x] Changelog
- [x] Docstrings

mfocko committed:
Switch pytest integration to hook

Switch the pytest integration to the hook instead of wrapping the runner
manually.

Fixes #152

Signed-off-by: Matej Focko <mfocko@redhat.com>

ollipa commented:
Thank you for the fix! The tests seem to fail for Pytest 5 but I think we should drop support for Pytest 5 in the next release along with Python 3.7.

ollipa committed:
Update test configuration and CI

ollipa committed:
Re-format files with black 2024 style

ollipa committed:
Fix linting issues

ollipa commented on src/flexmock/_pytest_plugin.py:
```diff
@@ -0,0 +1,26 @@
+# Hook flexmock into Pytest.
```
Can you change this comment to a docstring?

mfocko replied:
done

ollipa commented on src/flexmock/_pytest_plugin.py:
```diff
@@ -0,0 +1,26 @@
+def pytest_runtest_makereport(
+    item: Item,  # pylint: disable=unused-argument
+    call: CallInfo[None],
+) -> Generator[None, None, None]:
```
Can you add a function docstring here?

mfocko replied:
added

ollipa reviewed (approved):
Looks good and test suite passes! Can you update the CHANGELOG.md?

The linting should pass after you add docstring to the new pytest plugin file.

I made a PR to drop support for Python 3.7 and Pytest 5.x (https://github.com/flexmock/flexmock/pull/154). Feel free to rebase your changes onto that.

It seems that pytest yanked version 8.1.0 so anyone testing this branch should install pytest 8.1.0 explicitly.

ollipa committed:
Lock sphinx version to <7.0.0

Sphinx v7 dropped setuptools support and we
still use it.

mfocko committed:
Test against the latest pytest

8.1.1 at the time

Signed-off-by: Matej Focko <mfocko@redhat.com>

mfocko committed:
Download the latest specfile for Copr build and TF

When running Copr (RPM) build and Testing Farm, download the latest
version of the specfile from the Fedora downstream rather than using an
outdated specfile that is present in the repository.

Also update the config to match the latest state of the Packit Service.

Signed-off-by: Matej Focko <mfocko@redhat.com>

mfocko commented:
Sorry about the trouble with the pylint, I'm getting kinda inconsistent results with the `make` command :eyes: locally it fails on mypy already (untyped decorator), also pylint produces way more warnings (mostly newer type aliases for union/optional/etc.)

ollipa commented:
> Sorry about the trouble with the pylint, I'm getting kinda inconsistent results with the `make` command 👀 locally it fails on mypy already (untyped decorator), also pylint produces way more warnings (mostly newer type aliases for union/optional/etc.)

I think it is just because you are using newer Python version locally. The CI uses the minimum supported Python version.

Thank your work on this. I'll merge this PR soon and create a new release.
